Refactor: modernize softtakeover code

Swiftb0y opened the pull request:
* refactor mixxx::Time to be a std::chrono-compatible clock
* eliminate unnecessary allocation in `EffectKnobParameterSlot`
* remove unnecessary allocation in `SoftTakeoverTest*`
* eliminate checks in `SoftTakeoverCtrl::enable` by encoding constraints
  in the type system
* deduplicate common code in `SoftTakeoverTest`. Now bundled as
  `SoftTakeoverTestWithValue`
* remove unnecessary allocation in `SoftTakeoverCtrl`
* many more miscellaneous improvements to `SoftTakeover(-Ctrl)`.

I'm trying to add add a follow up commit that fixes the currently disabled test cases but it turns out the Truth table documented doesn't really match the unit tests (especially the currently enabled edge-case tests). So I've decided to not touch that logic for now. This is already a substantial improvement IMO.

Swiftb0y committed:
refactor: `mixxx::Time` to be a std::chrono-compatible clock

Swiftb0y committed:
refactor: modernize Softakeover

* eliminate unnecessary allocation in `EffectKnobParameterSlot`
* remove unnecessary allocation in `SoftTakeoverTest*`
* eliminate checks in `SoftTakeoverCtrl::enable` by encoding constraints
  in the type system
* deduplicate common code in `SoftTakeoverTest`. Now bundled as
  `SoftTakeoverTestWithValue`
* remove unnecessary allocation in `SoftTakeoverCtrl`
* many more miscellaneous improvements to `SoftTakeover(-Ctrl)`.

daschuer commented on src/controllers/softtakeover.h:
```diff
@@ -1,56 +1,97 @@
+    }
     // Check to see if the new value for the Control should be ignored
-    bool ignore(ControlObject* control, double newMidiParameter);
+    bool ignore(ControlObject* control, double newParameter) {
```
Please use p prefix for pointers.

daschuer reviewed (changes requested):
Thank you. I have added some comments for improvements.

Swiftb0y commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:
```diff
@@ -584,11 +587,12 @@ void ControllerScriptInterfaceLegacy::softTakeover(
+        if (!pControlPotmeter) {
+            return;
+        }
+        m_st.enable(gsl::not_null(pControlPotmeter));
```
> it exposes SoftTakeover internals to the outside without taking any benefits.

Can you elaborate? Why is the fact that softtakeover only works on `ControlPotMeter`s internal? IMO calling `st.enable(ControlObject*)` and then silently ignoring it because its not a `ControlPotMeter` is a much worse pattern...

>  At least disable() and enable() should take the same parameters.

so both `not_null<ControlPotMeter>`? I explicitly decided against that because it would've forced a null check and a cast onto the caller, both of which are unnecessary.

daschuer replied:
This function here is also invisible when setting the SoftTakeover flag. Currently enable and disable are only called here, that's why I think the conversion to `not_null<ControlPotMeter>` should be done above of set. 
... Or pass `set` es parameter and dispose `disable()`

Swiftb0y replied:
I'm afraid I can't follow you... 

> This function here is also invisible when setting the SoftTakeover flag

What do you mean with invisible? Invisible to whom?

> Currently enable and disable are only called here, that's why I think the conversion to not_null<ControlPotMeter> should be done above of set.

But that would incur the null check and the `qobject_cast` in the unnecessary disable case...

Swiftb0y replied:
Thinking about this more: if we would remove the `ControlObject*` completely from `SoftTakeover` and use `not_null<ControlObject*>` instead, that would allow `SoftTakeover` in a more type-safe fashion, but then we had the cast and the null check everywhere. I explicitly avoided that because I thought you might oppose it, but if you're in favor of that, I'm in favor too.

daschuer replied:
> But that would incur the null check and the qobject_cast in disable case even though its not needed...

Oh yes, that's not optimal. So we may keep it as it is in this PR.  

This started, because I was wondering why the ControlObject type check is now outside of the SoftTakeoverCtrl. Every mapping can have the softtakover flag set an it really dos not matter at which level in the call stack it is rejected, because it is invisible to the mapping author anyway. 

Original the pattern was "here is a CO with the flag set, please do soft takeover as good as you can". 
Now the caller needs to verify first if the the flag is set for the right CO type. We may have an issue if we create another type with Softtakover not inherited by ControlPotmeter.

Swiftb0y replied:
> Every mapping can have the softtakover flag set an it really dos not matter at which level in the call stack it is rejected, because it is invisible to the mapping author anyway.

Yes. Essentially because the ControlPotmeter concept doesn't really exist outside the C++ domain, in JS its just a collection of autogenerated COs.

> We may have an issue if we create another type with Softtakover not inherited by ControlPotmeter.

Yes, but at least now that is obvious since the API explicitly encodes that it is only expecting ControlPotmeters. So the more strongly-typed API is an improvement here IMO.

daschuer replied:
Ok, than keep it like this.

Swiftb0y commented on src/controllers/softtakeover.h:
```diff
@@ -1,56 +1,97 @@
+    // 3/128 units away from the current is enough to catch fast non-sequential moves
+    //  but not cause an audibly noticeable jump, determined experimentally with
+    //  slow-refresh controllers.
+    // TODO (XXX): Expose this to the controller mapping environment?
```
https://github.com/mixxxdj/mixxx/blob/6e8f8c522ac2409e805ae761a20e02bf4ad468ca/res/controllers/Numark-NS6II-scripts.js#L235-L246

To be fair, a `SoftTakeveover::isIgnoring` would be a better solution here since it doesn't expose as much.

daschuer replied:
Yes, do you have interest to add that?

Swiftb0y replied:
I'll see

Swiftb0y commented on src/effects/effectknobparameterslot.h:
```diff
@@ -49,7 +49,7 @@ class EffectKnobParameterSlot : public EffectParameterSlotBase {
     }
 
-    SoftTakeover* m_pMetaknobSoftTakeover;
+    SoftTakeover m_metaknobSoftTakeover{};
```
why?
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-in-class-initializer

daschuer replied:
In this case `SoftTakeover` is automatically initialized with its default constructor, the {} is redundant. 

In general we have the rule to prefer the constructor member initializer list:
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#non-static-data-member-initializers

In addition this practice avoids a static inline constructor call and minimizes recompilation in case the initializer changes and redundant initializers.

I just noticed there are also some pending issues: 
https://github.com/isocpp/CppCoreGuidelines/issues/2213
https://github.com/isocpp/CppCoreGuidelines/issues/2032

The real issue in this case is that we may forget to initialize all members. I am afraid we have not set the compiler/linter to warn us about that. We should consider to set it up.

Swiftb0y replied:
> In this case `SoftTakeover` is automatically initialized with its default constructor, the {} is redundant.

I am aware that this applies for class types, it does not however apply to primitive types so I wanted to establish this style regardless of what type it is in an effort to simplify the code...

Swiftb0y replied:
from what I could tell, `-Weffc++` does the trick. Unfortunately it flags lots of code, so enabling that will have to be a separate PR.

Swiftb0y commented on src/controllers/softtakeover.h:
```diff
@@ -1,56 +1,97 @@
     // Ignore the next supplied parameter
-    void ignoreNext(ControlObject* control);
+    void ignoreNext(ControlObject* control) {
+        auto coIt = m_softTakeoverHash.find(control);
+        if (coIt == m_softTakeoverHash.end()) {
+            return;
+        }
+
+        coIt->second.ignoreNext();
```
wbout this?
```suggestion
        auto coIt = m_softTakeoverHash.find(control);
        if (coIt == m_softTakeoverHash.end()) {
            return;
        }
        auto [key, softTakeoverInstance] = *coIt;
        softTakeoverInstance.ignoreNext();
```

daschuer replied:
This looks like a copied value, is it? What is the downside of the wrapper that returns a pointer? Normally we avoid non const references.

Swiftb0y replied:
> This looks like a copied value, is it?

Whoops, yes. make it `auto&` then.

> Normally we avoid non const references.

Does not apply here IMO. Or can you link me to the rationale behind this again?

> What is the downside of the wrapper that returns a pointer?

you force an unnecessarily unsafe API via the pointer that is being returned. You also introduce an extra also unnecessary nullcheck...

I also disagree that the anonymous pair hinders legibility much because thats just an integral part of `std::unordered_map`... Yes its ugly in the standard that they can't return a struct with named values, but thats how it is in the standard and I think its reasonable to expect to get used to it...

daschuer replied:
Here is our rule: 
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#non-const-references

The "Whoops" here confirms that the pointer syntax is a reasonable choice for cases like here. 
The idea with the wrapper function helps to de-duplicate the code. A pointer with a null check is fully save and when we inline the wrapper as well it will disappear in the machine code. 

However if you prefer to keep a single function this is would be OK for me.

This one fully matches our guideline:
```suggestion
        auto it = m_softTakeoverHash.find(control);
        if (it == m_softTakeoverHash.end()) {
            return;
        }
        SoftTakeover* pSt = &(it->second);
        pSt->second.ignoreNext();
```

This one is kind of OK after our discussion
```suggestion
        auto it = m_softTakeoverHash.find(control);
        if (it == m_softTakeoverHash.end()) {
            return;
        }
        auto& [key, refSoftTakeover] = *it;
        refSoftTakeover.ignoreNext(); 
```

Swiftb0y replied:
I prefer the latter one as I don't see the purpose of taking a pointer (`&(it->second)`) just to dereference it again immediately after (`pSt->second`).

Swiftb0y replied:
> Here is our rule:
> https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#non-const-references
> 
> The "Whoops" here confirms that the pointer syntax is a reasonable choice for cases like here.

I still don't agree with it. The rule seems to contrived and arbitrary... I will open a discussion on zulip.

Swiftb0y committed:
feat: add `engine.softTakeoverIsIgnoring` Javascript API

This is for mappings to determine whether setting a value would
result in actual changes to it or get it ignored because of
SoftTakeover and provide visual (LED) feedback for it.

Swiftb0y commented:
I took the liberty to mark all comments I have addressed in code as resolved so we can focus on clearing up the remaining questions.

daschuer commented on src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:
```diff
@@ -609,6 +609,16 @@ void ControllerScriptInterfaceLegacy::softTakeoverIgnoreNextValue(
+    if (!pControl) {
+        return false;
+    }
+    return m_st.ignore(pControl, parameter);
```
I think you want to implement a a dry run function, right? Ignore() is state full, updates the timestamp and the previous value so that a subsequent set will be never ignored.

Swiftb0y replied:
whoops, yeah, you're right.

daschuer commented on src/controllers/softtakeover.h:
```diff
@@ -1,56 +1,98 @@
     // Check to see if the new value for the Control should be ignored
-    bool ignore(ControlObject* control, double newMidiParameter);
+    bool ignore(ControlObject* pControl, double newParameter) {
+        auto coIt = m_softTakeoverHash.find(pControl);
+        if (coIt == m_softTakeoverHash.end()) {
+            return false;
+        }
+        VERIFY_OR_DEBUG_ASSERT(pControl) {
+            return false;
+        }
+        auto& [coKey, softTakeoverInstance] = *coIt;
+        return softTakeoverInstance.ignore(*pControl, newParameter);
```
Without explicit types we need to make sure the var names are not misleading.
How about that?  
```suggestion
        auto it = m_softTakeoverHash.find(pControl);
        if (it == m_softTakeoverHash.end()) {
            return false;
        }
        VERIFY_OR_DEBUG_ASSERT(pControl) {
            return false;
        }
        auto& [key, refSoftTakeover] = *it;
        return refSoftTakeover.ignore(*pControl, newParameter);
```

daschuer commented on src/controllers/softtakeover.h:
```diff
@@ -1,56 +1,98 @@
     // Ignore the next supplied parameter
-    void ignoreNext(ControlObject* control);
+    void ignoreNext(ControlObject* pControl) {
+        auto coIt = m_softTakeoverHash.find(pControl);
```
The same like above.

daschuer reviewed (approved):
LGTM. Thank you. Please ping me once it is ready for merge.

Swiftb0y commented:
@daschuer done. verified locally that each commit still builds and test pass.

Swiftb0y commented:
also added a test for `willIgnore` and amended. just fyi

daschuer commented:
Thank you.
